src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -652,5 +722,30 @@ let add_watch t path =
   | Inotify inotify -> (
     try Ok (Inotify_lib.add inotify (Path.to_string path))
     with Unix.Unix_error (ENOENT, _, _) -> Error `Does_not_exist)
+  | Fswatch_win fswatch -> (
+    match path with
+    | In_build_dir _ ->
+      Code_error.raise "attempted to watch a directory in build" []
+    | Path.In_source_tree _ -> Ok ()
+    | External ext -> (
+      let ext =
+        let rec loop p =
+          if Path.is_directory (Path.external_ p) then Some ext
+          else
+            match Path.External.parent p with
+            | None ->
+              User_warning.emit
+                [ Pp.textf "Refusing to watch %s" (Path.External.to_string ext)
+                ];
+              None
+            | Some ext -> loop ext
+        in
+        loop ext
```

Comment:
Share this code with the fsevents watcher somehow?

Reply:
Fixed, thanks.